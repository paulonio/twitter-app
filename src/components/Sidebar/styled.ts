import { styled } from 'styled-components';

import SearchIcon from '@src/icons/SearchIcon';

export const FixedWrapper = styled.div`
  width: inherit;
  position: fixed;
  z-index: 10;
`;

export const Wrapper = styled.aside`
  width: 373px;
  margin-top: 20px;

  @media screen and (max-width: 1480px) {
    display: none;
  }
`;

export const InputWrapper = styled.div`
  width: 100%;
  max-width: 373px;
  position: relative;
`;

export const Label = styled.label`
  position: relative;
  display: block;
  border-radius: 50px;
`;

export const SearchIconcc = styled.img.attrs({
  src: '/assets/icons/Search.svg',
  alt: 'Search icon',
})`
  position: absolute;
  width: 24px;
  height: 24px;
  top: 50%;
  left: 20px;
  transform: translateY(-50%);
`;

export const Search = styled(SearchIcon)`
  position: absolute;
  width: 24px;
  height: 24px;
  top: 50%;
  left: 20px;
  transform: translateY(-50%);
`;

export const SearchField = styled.input.attrs({ placeholder: 'Search Twitter' })`
  width: 100%;
  color: ${({ theme }) => theme.colors.textInput};
  background-color: ${({ theme }) => theme.colors.backgroundInput};
  padding: 15px 0;
  padding-left: 64px;
  font-size: 18px;
  line-height: 24px;
  font-weight: 400;
  border: none;
  border-radius: 50px;
`;

export const PostsWrapper = styled.div`
  width: 373px;
  margin-top: 31px;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  border-radius: 10px;
  overflow: hidden;
`;

export const Post = styled.div`
  width: 33%;
  height: 88px;
`;

export const PostImage = styled.img`
  width: 100%;
  height: 100%;
  background-color: black;
`;

export const SearchResults = styled.div`
  width: 100%;
  position: absolute;
  top: calc(100% + 10px);
  left: 0;
  padding: 10px;
  background-color: ${({ theme }) => theme.colors.backgroundInput};
  border-radius: 8px;
  &:hover {
    cursor: pointer;
  }
`;

export const SearchResult = styled.p`
  color: #5c6c79;
`;

export const UsersWrapper = styled.div`
  width: 373px;
  padding: 20px 15px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
`;

export const UserRow = styled.div`
  width: 100%;
  margin-top: 25px;
  margin-bottom: 30px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const UsersTitle = styled.h3`
  font-size: 24px;
  line-height: 32px;
  font-weight: 700;
  color: ${({ theme }) => theme.colors.text};
`;

export const ButtonWrapper = styled.div`
  width: 100%;
  max-width: 92px;
`;
