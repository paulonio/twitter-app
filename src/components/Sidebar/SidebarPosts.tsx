import React from 'react';

import { POSTS_DATA } from '@src/constants/constants';

import { Post, PostImage, PostsWrapper } from './styled';

const SidebarPosts = () => {
  return (
    <PostsWrapper>
      {POSTS_DATA.map(({ id, src }) => (
        <Post key={id}>
          <PostImage src={src} alt="Post" />
        </Post>
      ))}
    </PostsWrapper>
  );
};

export default SidebarPosts;
