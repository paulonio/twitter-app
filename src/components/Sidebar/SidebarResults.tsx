import React, { FC, MouseEvent } from 'react';
import { useNavigate } from 'react-router-dom';

import MenuUser from '@src/components/Menu/MenuUser';

import { TweetType } from '@store/slices/tweetSlice';
import type { UserType } from '@store/slices/authSlice';

import { isUserArray } from '@utils/utils';

import { SearchResult, SearchResults } from './styled';

interface ResultsProps {
  data: TweetType[] | UserType[];
}

const SidebarResults: FC<ResultsProps> = ({ data }) => {
  const navigate = useNavigate();

  const handleNavigateToTweetPage = (e: MouseEvent<HTMLDivElement>) => {
    const id = e.currentTarget.getAttribute('data-id');
    navigate(`/feed/${id}`);
  };

  if (isUserArray(data)) {
    return (
      <SearchResults>
        {data.length > 0 && data.map((user) => <MenuUser key={user.uid} user={user} />)}
      </SearchResults>
    );
  }

  return (
    <SearchResults>
      {data.map(({ tweet, id }) => (
        <SearchResult key={id} onClick={handleNavigateToTweetPage} data-id={id}>
          {tweet}
        </SearchResult>
      ))}
    </SearchResults>
  );
};

export default SidebarResults;
