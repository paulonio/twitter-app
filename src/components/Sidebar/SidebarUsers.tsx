import React from 'react';
import { useSelector } from 'react-redux';

import MenuUser from '@src/components/Menu/MenuUser';
import { StyledButton } from '@components/Button/styled';

import type { StoreType } from '@store/index.ts';
import { UserType } from '@store/slices/authSlice';

import { UsersTitle, UsersWrapper, UserRow, ButtonWrapper } from './styled';

const SidebarUsers = () => {
  const user = useSelector<StoreType, UserType | null>((state) => state.auth.user);

  return (
    <UsersWrapper>
      <UsersTitle>You might like</UsersTitle>
      <UserRow>
        {user && <MenuUser user={user} />}
        <ButtonWrapper>
          <StyledButton $buttonType="secondary">Follow</StyledButton>
        </ButtonWrapper>
      </UserRow>
      <StyledButton $buttonType="link">Show more</StyledButton>
    </UsersWrapper>
  );
};

export default SidebarUsers;
