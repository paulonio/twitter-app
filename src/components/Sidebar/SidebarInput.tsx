import React, { ChangeEvent, FC } from 'react';

import SidebarResults from '@src/components/Sidebar/SidebarResults';

import type { TweetType } from '@store/slices/tweetSlice';
import type { UserType } from '@store/slices/authSlice';

import { Label, SearchField, SearchIcon, InputWrapper } from './styled';

interface SidebarInputProps {
  value: string;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  data: TweetType[] | UserType[];
}

const SidebarInput: FC<SidebarInputProps> = ({ value, onChange, data }) => {
  return (
    <InputWrapper>
      <Label>
        <SearchIcon />
        <SearchField value={value} onChange={onChange} data-testid="search" />
        {data && value && <SidebarResults data={data} />}
      </Label>
    </InputWrapper>
  );
};

export default SidebarInput;
