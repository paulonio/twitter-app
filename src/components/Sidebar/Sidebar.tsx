import React, { FC } from 'react';

import SidebarInput from '@src/components/Sidebar/SidebarInput';
import SidebarPosts from '@src/components/Sidebar/SidebarPosts';
import SidebarUsers from '@src/components/Sidebar/SidebarUsers';

import { useSidebarHandlers } from '@src/hooks/useSidebarHandlers';
import { FixedWrapper, Wrapper } from './styled';

interface SidebarProps {
  type: 'profile' | 'feed';
}

const Sidebar: FC<SidebarProps> = ({ type }) => {
  const {
    usersFilterValue,
    tweetsFilterValue,
    filteredUsers,
    filteredTweets,
    handleChangeUsersValue,
    handleChangeTweetsValue,
  } = useSidebarHandlers();

  return (
    <Wrapper>
      <FixedWrapper>
        {type === 'profile' ? (
          <SidebarInput
            value={usersFilterValue}
            onChange={handleChangeUsersValue}
            data={filteredUsers}
          />
        ) : (
          <SidebarInput
            value={tweetsFilterValue}
            onChange={handleChangeTweetsValue}
            data={filteredTweets}
          />
        )}
        <SidebarPosts />
        <SidebarUsers />
      </FixedWrapper>
    </Wrapper>
  );
};

export default Sidebar;
