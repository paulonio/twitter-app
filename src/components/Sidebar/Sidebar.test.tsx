import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import Sidebar from '@components/Sidebar/Sidebar';
import CustomProvider from '@components/EditUser/testUtils';

describe('Profile', () => {
  it('should render profile component', () => {
    const { container } = render(
      <CustomProvider>
        <Sidebar type="profile" />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should change search correctly', () => {
    render(
      <CustomProvider>
        <Sidebar type="profile" />
      </CustomProvider>
    );

    const search: HTMLInputElement = screen.getByTestId('search');
    fireEvent.change(search, { target: { value: 'Test' } });
    expect(search.value).toEqual('Test');
  });
});
