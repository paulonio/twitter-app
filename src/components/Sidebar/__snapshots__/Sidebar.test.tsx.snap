// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile should render profile component 1`] = `
<div>
  <aside
    class="sc-burPHO dEhgfc"
  >
    <div
      class="sc-jySWPz jApgcW"
    >
      <div
        class="sc-hlwzUd hloTfY"
      >
        <label
          class="sc-kwrclo eSVPPH"
        >
          <svg
            class="sc-gaYwgq ciCPNQ"
            fill="currentColor"
            focusable="false"
            height="24"
            role="img"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M21.53 20.47L17.87 16.81C19.195 15.24 20 13.214 20 11C20 6.03 15.97 2 11 2C6.03 2 2 6.03 2 11C2 15.97 6.03 20 11 20C13.215 20 15.24 19.196 16.808 17.87L20.468 21.53C20.615 21.676 20.808 21.75 20.998 21.75C21.188 21.75 21.383 21.677 21.528 21.53C21.823 21.237 21.823 20.763 21.53 20.47ZM3.5 11C3.5 6.865 6.865 3.5 11 3.5C15.135 3.5 18.5 6.865 18.5 11C18.5 15.135 15.135 18.5 11 18.5C6.865 18.5 3.5 15.135 3.5 11Z"
              fill="#5C6C79"
            />
          </svg>
          <input
            class="sc-goxvPd kPxVJK"
            data-testid="search"
            placeholder="Search Twitter"
            value=""
          />
        </label>
      </div>
      <div
        class="sc-jzrYgo evmPBv"
      >
        <div
          class="sc-cSiKfk hgmRfT"
        >
          <img
            alt="Post"
            class="sc-XDmMx ktECaL"
            src="/assets/images/post-1.jpg"
          />
        </div>
        <div
          class="sc-cSiKfk hgmRfT"
        >
          <img
            alt="Post"
            class="sc-XDmMx ktECaL"
            src="/assets/images/post-2.jpg"
          />
        </div>
        <div
          class="sc-cSiKfk hgmRfT"
        >
          <img
            alt="Post"
            class="sc-XDmMx ktECaL"
            src="/assets/images/post-3.jpg"
          />
        </div>
        <div
          class="sc-cSiKfk hgmRfT"
        >
          <img
            alt="Post"
            class="sc-XDmMx ktECaL"
            src="/assets/images/post-4.jpg"
          />
        </div>
        <div
          class="sc-cSiKfk hgmRfT"
        >
          <img
            alt="Post"
            class="sc-XDmMx ktECaL"
            src="/assets/images/post-5.jpg"
          />
        </div>
        <div
          class="sc-cSiKfk hgmRfT"
        >
          <img
            alt="Post"
            class="sc-XDmMx ktECaL"
            src="/assets/images/post-6.jpg"
          />
        </div>
      </div>
      <div
        class="sc-hyPqDM bXpQbo"
      >
        <h3
          class="sc-cArpXK kSfmxI"
        >
          You might like
        </h3>
        <div
          class="sc-fEjTkZ VsgPP"
        >
          <div
            class="sc-FLSEX htyKW"
          >
            <button
              class="sc-gBQmyM ithDjM"
            >
              Follow
            </button>
          </div>
        </div>
        <button
          class="sc-gBQmyM fdWkEv"
        >
          Show more
        </button>
      </div>
    </div>
  </aside>
</div>
`;
